cmd/fs-v1-append-parts.go

```diff
@@ -0,0 +1,170 @@
+package cmd
+
+import (
+	"errors"
+	"io"
+	"reflect"
+	"sync"
+	"time"
+)
+
+// Error sent by appendRoutine when there are holes in parts.
+// For ex. let's say client uploads part-2 before part-1 in which case we
+// can not append and have to wait till part-1 is uploaded. Hence we return
+// this error. Currently this error is not used in the caller.
+var errPartsMissing = errors.New("required parts missing")
+
+// Error sent when appendRoutine has waited long enough and timedout.
+var errAppendRoutineTimeout = errors.New("timeout")
+
+// Timeout value for the appendRoutine go-routine.
+var appendRoutineTimeout = 24 * 60 * 60 * time.Second
+
+// Holds a map of uploadID->appendRoutine
+type appendPartFS struct {
+	appendRoutineMap map[string]appendRoutineInfo
+	mu               sync.Mutex
+}
+
+// Input to the appendRoutine
+type appendRoutineInput struct {
+	meta  fsMetaV1   // list of parts that need to be appended
+	errCh chan error // error sent by appendRoutine
+}
+
+// Identifies an appendRoutine.
+type appendRoutineInfo struct {
+	inputCh   chan appendRoutineInput
+	timeoutCh chan struct{} // closed by appendRoutine when it timesout
+	endCh     chan struct{} // closed after complete/abort of upload to end the appendRoutine
+}
+
+// Called after a part is uploaded so that it can be appended in the background.
+func (m *appendPartFS) processPart(disk StorageAPI, bucket, object, uploadID string, meta fsMetaV1) chan error {
+	m.mu.Lock()
+	info, ok := m.appendRoutineMap[uploadID]
+	if !ok {
+		// Corresponding appendRoutine was not found, create a new one. Would happen when the first
+		// part of a multipart upload is uploaded.
+		inputCh := make(chan appendRoutineInput)
+		timeoutCh := make(chan struct{})
+		endCh := make(chan struct{})
+
+		info = appendRoutineInfo{inputCh, timeoutCh, endCh}
+		m.appendRoutineMap[uploadID] = info
+
+		go m.appendRoutine(disk, bucket, object, uploadID, info)
+	}
+	m.mu.Unlock()
+	errCh := make(chan error)
+	go func() {
+		// send input in a goroutine as send on the inputCh can block if appendRoutine
+		// is busy appending a part.
+		select {
+		case <-info.timeoutCh:
+			// This is to handle a rare race condition where we found info in m.appendRoutineMap
+			// but soon after that appendRoutine timedouted out.
+			errCh <- errAppendRoutineTimeout
+		case info.inputCh <- appendRoutineInput{meta, errCh}:
+		}
+	}()
+	return errCh
+}
+
+// Called after complete-multipart-upload or abort-multipart-upload so that the appendRoutine is not left dangling.
+func (m *appendPartFS) endAppendRoutine(uploadID string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	info, ok := m.appendRoutineMap[uploadID]
+	if !ok {
+		return
+	}
+	delete(m.appendRoutineMap, uploadID)
+	close(info.endCh)
+}
+
+// The go-routine that appends the parts in the background.
+func (m *appendPartFS) appendRoutine(disk StorageAPI, bucket, object, uploadID string, info appendRoutineInfo) {
```

Comment:
since it's a go-routine that runs in the background, i named it like a noun i.e it's a goroutine that's responsible for appending so the go-routine is called `appendRoutine` if it was a function then appendInRoutine would be better as it's a verb